[Index] Force indexing of system modules to read only from swiftinterfaces

xymus opened the pull request:
Indexing system modules while building is a crash prone phase because it tries to read everything from swiftmodules in the SDK. We saw it crash often on reading implementation details relying on implementation-only imported types.

To rule out all of these issues, let's force indexing system modules to use only swiftinterfaces in the SDK and ignores the adjacent swiftmodule files. In the future, we want a similar behavior for all compilation, not just indexing, at that time we can revert this change.

rdar://100644036

xymus committed:
[Frontend] Intro global control to force loading from swiftinterface

Intro ASTContext::setIgnoreAdjacentModules to change module loading to
accept load only resilient modules from their swiftinterfaces, ignoring
the adjacent module and any silencing swiftinterfaces errors.

xymus committed:
[Index] Force indexing from the swiftinterfaces of system modules

Use setIgnoreAdjacentModules before indexing any resilient system module
to force indexing to read only from resilient modules.

rdar://100644036

xymus committed:
[Test] Use a clean module cache for more tests

Use a clean module cache in tests printing path to the
swiftmodule/swiftinterface used as it may change depending on previous
tests.

xymus commented:
@swift-ci Please smoke test

benlangmuir commented:
Am I correct in understanding that we will still load prebuilt binary modules from the sdk (in the separate prebuilt directory), just not adjacent ones?

benlangmuir commented on lib/Index/IndexRecord.cpp:
```diff
@@ -495,6 +495,22 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+    module->getASTContext().setIgnoreAdjacentModules(true);
+
+    ImportPath::Module::Builder builder(module->getName());
+    auto reloadedModule = module->getASTContext().getModule(builder.get());
```
I'm concerned about the cost of re-reading the module here.  Even if nothing is later deserialized from the module, my understanding is that loading the module has a non-trivial cost (this was important in code-completion for example).  Indexing-while-building is enabled by default in all debug builds, so this would have a wide impact.

Can we move the reload of the module below the `parentUnitWriter.isUnitUpToDateForOutputFile` check below? Then we would only incur this cost when we will actually need to index the given module. We would still need the updated module filename for that check, but not to actually read the file.

xymus replied:
Good point, I moved the new logic after the "Up to date check". I also tweaked it in the hope the we still write down something when the swiftinterface fails to build, so the next indexing phase won't attempt to rebuild it again.

I'm not sure how to test superfluous reindexing though, please let me know if you have an idea.

bnbarham replied:
> I'm not sure how to test superfluous reindexing though, please let me know if you have an idea.

We could add a remark (`indexing-system-module`), add `-Rindexing-system-module` to the test, and then check the output of the second compile doesn't contain it?

xymus replied:
Good idea, I added this remark and used it in the test. It shows on attempts to index, whenever we write a file, and specifies if the content is fixed because of a broken swiftinterface.

xymus committed:
[Test] Avoid unexpected deserialization failure in skip-loaded-internal.swift

xymus commented:
Pushing a new version to run test. I still have some offline comments to look into.

@swift-ci Please smoke test

xymus committed:
[Index] Move reloading from swiftinterface after the "up to date" check

xymus committed:
[Index] Skip indexing modules failing to build but write it down

xymus commented:
> Am I correct in understanding that we will still load prebuilt binary modules from the sdk (in the separate prebuilt directory), just not adjacent ones?

That's correct, only the adjacent modules are ignored.

xymus commented:
@swift-ci Please smoke test

bnbarham commented on lib/Index/IndexRecord.cpp:
```diff
@@ -513,14 +514,33 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
   if (*isUptodateOpt)
     return false;
 
+  // Reload resilient modules from swiftinterface to avoid indexing
```
Is it possible to get the name of the reloaded module before actually loading it? We'll be using the `.swiftmodule` name now which is a bit weird, but maybe that's fine 🤔

xymus replied:
With the last change I think that's ok, we set aside `filename` before reloading.

bnbarham commented on lib/Index/IndexRecord.cpp:
```diff
@@ -513,14 +514,33 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
 
-  if (!module->isStdlibModule()) {
+  if (skipIndexingModule) {
+    // Don't add anything to records.
```
It might be worth adding a comment on the imported modules section below explaining that we believe that's still safe enough to do.

xymus replied:
I've added that we keep writing the file to avoid rebuilding on broken swiftinterfaces.

bnbarham commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -444,7 +445,8 @@ std::error_code ImplicitSerializedModuleLoader::findModuleFilesInDirectory(
 
-  if (LoadMode == ModuleLoadingMode::OnlyInterface)
+  if (LoadMode == ModuleLoadingMode::OnlyInterface ||
+      (Ctx.IgnoreAdjacentModules && ModuleID.Item.str() != STDLIB_NAME))
```
So we still read from the module for the stdlib?

xymus replied:
Yes, we still allow the stdlib because too many tests rely on it using the swiftmodule built from full source. We'll likely want to change that behavior in the future as using the stdlib rebuilt from the swiftinterface provided results closer to what 3rd party devs see.

xymus commented on lib/Index/IndexRecord.cpp:
```diff
@@ -495,6 +495,22 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+  // Reload resilient modules from swiftinterface to avoid indexing
+  // internal details.
+  if (module->getResilienceStrategy() == ResilienceStrategy::Resilient) {
+    module->getASTContext().setIgnoreAdjacentModules(true);
```
Yeah, this is an attempt at having this mode turn on only in case there are resilient dependencies. We don't need to clear the cache otherwise. It's not ideal but it may provide some small gains until we have all the compile ignore the adjacent swiftmodule file.

xymus committed:
[Index] Remark on indexing system modules and use it in tests

xymus commented:
@swift-ci Please smoke test

bnbarham commented on lib/Index/IndexRecord.cpp:
```diff
@@ -513,14 +514,40 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+                   filename, skipIndexingModule);
+  }
+
   // FIXME: Would be useful for testing if swift had clang's -Rremark system so
```
FWIW you just implemented this FIXME :P

xymus replied:
Ah, right, I deleted the FIXME. This work was needed apparently!

bnbarham commented on lib/Serialization/SerializedModuleLoader.cpp:
```diff
@@ -444,7 +445,9 @@ std::error_code ImplicitSerializedModuleLoader::findModuleFilesInDirectory(
-  if (LoadMode == ModuleLoadingMode::OnlyInterface)
+  if (LoadMode == ModuleLoadingMode::OnlyInterface ||
+      (Ctx.IgnoreAdjacentModules && ModuleID.Item.str() != STDLIB_NAME &&
+       ModuleID.Item.str() !=SWIFT_STRING_PROCESSING_NAME))
```
Nitpick: space after `!=`. But if there's not any other changes I'd just leave it.

bnbarham commented on test/Index/index_system_modules_swiftinterfaces.swift:
```diff
@@ -0,0 +1,94 @@
+// RUN:     -Rindexing-system-module \
+// RUN:     %t/Client.swift \
+// RUN:     2>&1 | %FileCheck -check-prefix=BROKEN-BUILD-2 --allow-empty %s
+// BROKEN-BUILD-2-NOT: indexing system module
```
Nice!

xymus replied:
It's working nicely!

bnbarham reviewed (approved):
Thanks Alexis! This seems reasonable to me, especially if ignoring the adjacent swiftmodule is something we're moving towards for the build anyway.

With the change to not attempt to reload broken modules, the only slowdown now will be the initial reload-from-interface for each module 👍. I think that's a decent tradeoff to avoid all the crashers we've been seeing.

xymus committed:
[Index] Apply small comments from PR review

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test macOS

xymus commented:
@swift-ci Please smoke test macOS

xymus committed:
[Sema|Index] Ignore the cache when loading a system module for indexing

xymus committed:
[Index] Do not reload the stdlib when indexing system modules

xymus commented:
Tweaked how this interacts with the in-memory cached modules. Instead of dropping the module cache it nows ignores the cache only to reload the target system module to index, and it doesn't attempt to reload the stdlib as it can break existing references to it.

@swift-ci Please smoke test

bnbarham commented on lib/AST/ASTContext.cpp:
```diff
@@ -2441,11 +2445,12 @@ bool ASTContext::canImportModule(ImportPath::Module ModuleName,
 
 ModuleDecl *
-ASTContext::getModule(ImportPath::Module ModulePath) {
+ASTContext::getModule(ImportPath::Module ModulePath, bool AllowMemoryCached) {
```
Should we not add the module to the loaded modules when `AllowMemoryCached` is set? Ie. `SerializedModuleLoader.cpp` will call `Ctx.addLoadedModule(M);` after loading the module and that will replace the already loaded one. But that means that if index module A and then module B, which depends on A, we would then have the loaded-from-interface module A when indexing B.

Or does not adding the module cause issues later, eg. something could be expecting to find that module in the loaded modules map and then getting the old B when we're indexing the loaded-from-interface B?

Mostly I'm concerned that we're not expecting this type of thing to happen in the compiler, so it's possible something is caching results that will cause problems here. I don't *know* if that's true. Though the comment about reloading the stdlib breaking references would point to something like this.

xymus replied:
I agree with you about the potential problem here. If the current approach works, I'll look into adding what you suggest, at least for completeness purposes and it may be more important when indexing larger projects.

xymus commented:
@swift-ci Please smoke test macOS

xymus committed:
[Serialization] Add control over adding a loaded module to the in-memory cache

xymus committed:
[Index] Don't reload modules already built from swiftinterface

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus commented:
https://github.com/apple/swift-package-manager/pull/5851

@swift-ci Please smoke test macOS

xymus commented:
Extracted some of the new serialization logic to a different PR and rebased on it.

https://github.com/apple/swift-package-manager/pull/5851

@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test Linux

benlangmuir commented on include/swift/AST/Decl.h:
```diff
@@ -632,6 +632,10 @@ class alignas(1 << DeclAlignInBits) Decl : public ASTAllocated<Decl> {
     RawResilienceStrategy : 1,
 
+    /// Whether the module was rebuilt from a module interface instead of being
+    /// build from the full source.
```
typo: build -> built

benlangmuir commented on include/swift/AST/Module.h:
```diff
@@ -544,6 +544,15 @@ class ModuleDecl
   }
 
+  /// Returns true if the module was rebuilt from a module interface instead
+  /// of being build from the full source.
```
Typo: build -> built

benlangmuir commented on include/swift/ClangImporter/ClangImporter.h:
```diff
@@ -220,11 +220,15 @@ class ClangImporter final : public ClangModuleLoader {
   /// the dotted module name to load, e.g., AppKit.NSWindow.
   ///
+  /// \param AllowMemoryCache Enables preserving the loaded module in the
+  /// in-memory cache for the next loading attempt.
```
The implementation appears to ignore this parameter, so I think we should change this particular `ClangImporter::loadModule` doc comment to reflect reality.

xymus replied:
Good point, I updated it.

benlangmuir commented on lib/Frontend/ModuleInterfaceBuilder.cpp:
```diff
@@ -81,12 +83,16 @@ struct ErrorDowngradeConsumerRAII: DiagnosticConsumer {
     }
     Diag.removeConsumer(*this);
+    if (Silence)
+      Diag.resetHadAnyError();
```
Maybe I'm missing something, but wouldn't this cause us to ignore previous (unrelated) errors for the purposes of checking whether an error has occurred?  Maybe we can do a save/restore of the error state instead?  Or if this cannot happen, maybe we could assert there haven't been any errors yet?

benlangmuir commented on lib/Frontend/ModuleInterfaceBuilder.cpp:
```diff
@@ -322,13 +329,23 @@ bool ImplicitModuleInterfaceBuilder::buildSwiftModuleInternal(
+          };
+
+          if (silenceInterfaceDiagnostics) {
+            ErrorDowngradeConsumerRAII R(*diags, /*Silence=*/true);
```
Feel free to ignore this suggestion, but it might be nicer to avoid the lambda:
```c++
Optional<ErrorDowngradeConsumerRAII> R;
if (silenceInterfaceDiagnostics)
  R.emplace(*diags, /*Silence=*/true);
...
```

xymus replied:
Thank you for the suggestion, I ended up applying this logic with the local temporary DiagnosticEngine.

benlangmuir commented on lib/Frontend/ModuleInterfaceLoader.cpp:
```diff
@@ -677,7 +677,8 @@ class ModuleInterfaceLoaderImpl {
     // Should we attempt to load a swiftmodule adjacent to the swiftinterface?
-    bool shouldLoadAdjacentModule = true;
+    bool shouldLoadAdjacentModule = !ctx.IgnoreAdjacentModules ||
+      moduleName == STDLIB_NAME;
```
Could we have a comment here about why we except the stdlib?

xymus replied:
This wasn't needed anymore actually so I removed that commit.

benlangmuir commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -424,6 +424,15 @@ static bool buildModuleFromInterface(CompilerInstance &Instance) {
+      Invocation.getFrontendOptions().RequestedAction !=
+        FrontendOptions::ActionType::TypecheckModuleFromInterface &&
+      !Invocation.getFrontendOptions().ExplicitInterfaceBuild) {
+      IgnoreAdjacentModules = Instance.getASTContext().IgnoreAdjacentModules;
```
These conditions are a bit opaque to me - could we have a comment explaining why we don't honour the ASTContext flag in these cases?

xymus replied:
Those are cases when the outer compiler instance doesn't instantiate an ASTContext. I replaced this check with the more direct check of `Instance.hasASTContext`.

benlangmuir replied:
Nice, that makes it obvious what's going on. Thanks!

benlangmuir commented on lib/Serialization/ModuleFileSharedCore.h:
```diff
@@ -347,6 +347,10 @@ class ModuleFileSharedCore {
     unsigned ResilienceStrategy : 2;
 
+    /// Whether the module was rebuilt from a module interface instead of being
+    /// build from the full source.
```
Typo: build -> built

xymus commented:
```
Cannot contact i-09a88659ae06fbf33: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@3cbfe351:i-09a88659ae06fbf33": Remote call on i-09a88659ae06fbf33 failed. The channel is closing down or has closed down
```

@swift-ci Please smoke test Linux

xymus committed:
[Serialization] Minor improvements to doc from PR comments

xymus committed:
[ModuleInterface] Implement silencing errors on rebuild with a new engine

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test
